drivers/xen/gnttab.c

```diff
@@ -324,46 +331,150 @@ const char *gnttabop_error(int16_t status)
 	}
 }
 
-static int gnttab_init(void)
+static int map_grant_frames(unsigned int start_frame, unsigned int end_frame)
 {
-	grant_ref_t gref;
-	struct xen_add_to_physmap xatp;
+	int rc, i;
+	unsigned int nr_frames = end_frame - start_frame;
 	struct gnttab_setup_table setup;
-	xen_pfn_t frames[NR_GRANT_FRAMES];
-	int rc = 0, i;
-
-	/* Will be taken/given during gnt_refs allocation/release */
-	k_sem_init(&gnttab.sem, 0, NR_GRANT_ENTRIES - GNTTAB_NR_RESERVED_ENTRIES);
-
-	for (
-		gref = GNTTAB_NR_RESERVED_ENTRIES;
-		gref < NR_GRANT_ENTRIES;
-		gref++
-	    ) {
-		put_free_entry(gref);
+	xen_pfn_t *frames;
+
+	if (gnttab.nr_grant_frames + nr_frames > gnttab.max_grant_frames) {
+		LOG_ERR("Reached max number of Xen grant frames");
+		return -ENOMEM;
 	}
 
-	for (i = 0; i < NR_GRANT_FRAMES; i++) {
+	for (i = start_frame; i < end_frame; i++) {
+		struct xen_add_to_physmap xatp;
+
+		/* Stage 2 mapping for first frame */
 		xatp.domid = DOMID_SELF;
 		xatp.size = 0;
 		xatp.space = XENMAPSPACE_grant_table;
 		xatp.idx = i;
 		xatp.gpfn = xen_virt_to_gfn(Z_TOPLEVEL_ROM_NAME(grant_tables).phys_addr) + i;
 		rc = HYPERVISOR_memory_op(XENMEM_add_to_physmap, &xatp);
-		__ASSERT(!rc, "add_to_physmap failed; status = %d\n", rc);
+		if (rc) {
+			LOG_ERR("add_to_physmap failed; status = %d\n", rc);
+			return rc;
+		}
+	}
+
+	gnttab.nr_grant_frames += nr_frames;
+
+	frames = k_calloc(gnttab.nr_grant_frames, sizeof(*frames));
+	if (!frames) {
+		LOG_ERR("Failed to allocate memory for frames");
+		return -ENOMEM;
 	}
 
+	/* Frames are mapped, so we can setup grant table */
 	setup.dom = DOMID_SELF;
-	setup.nr_frames = NR_GRANT_FRAMES;
+	setup.nr_frames = gnttab.nr_grant_frames;
 	set_xen_guest_handle(setup.frame_list, frames);
 	rc = HYPERVISOR_grant_table_op(GNTTABOP_setup_table, &setup, 1);
-	__ASSERT((!rc) && (!setup.status), "Table setup failed; status = %s\n",
-		gnttabop_error(setup.status));
+	if (rc || setup.status) {
+		LOG_ERR("Table setup failed; status = %s", gnttabop_error(setup.status));
```

Comment:
I am curious about state of the gnttab after this. We mapped a frame but we couldn't setup it. What will happen during the next call to this function?

Also, should you setup `rc` here?

Reply:
> From Linux and Xen it seems we do not need to call GNTTABOP_setup_table for ARM, the return results are used only by X86.

It is supposed to be the common code. AFAIK, there are experiments with running Zephyr-xenvm on x86.

Reply:
> What will happen during the next call to this function?

I checked Linux driver - in same conditions it just calls `BUG_ON()` and continues execution. In our case, as @GrygiriiS said, this call is not doing much inside Xen, but interesting that it may both return negative errno in `rc` or zero `rc` with error in `setup.status`. Since we were able to add one more frame with `memory_op`, I think this is definitely not OK condition, so I'm now setting `rc` here also from `setup.status`

> From Linux and Xen it seems we do not need to call GNTTABOP_setup_table for ARM, the return results are used only by X86.

As I can see, `static int gnttab_map()` doing the same for both of them, only `map_frames()` callbacks looks different in platform specific implementation.

Reply:
> > What will happen during the next call to this function?
> 
> I checked Linux driver - in same conditions it just calls `BUG_ON()` and continues execution. In our case, as @GrygiriiS said, this call is not doing much inside Xen, but interesting that it may both return negative errno in `rc` or zero `rc` with error in `setup.status`. Since we were able to add one more frame with `memory_op`, I think this is definitely not OK condition, so I'm now setting `rc` here also from `setup.status`
> 
> > From Linux and Xen it seems we do not need to call GNTTABOP_setup_table for ARM, the return results are used only by X86.
> 
> As I can see, `static int gnttab_map()` doing the same for both of them, only `map_frames()` callbacks looks different in platform specific implementation.

If you check arch_gnttab_map_status() and arch_gnttab_map_shared() in Linux for ARM - they are NOP.
Hence there is NO proper implementation for Zephyr x86 and Zephyr XEN depends on ARMV8_A, I'd just drop this dead code and add comment why.

Reply:
> If you check arch_gnttab_map_status() and arch_gnttab_map_shared() in Linux for ARM - they are NOP.

Yeah, I mentioned it here:
> only map_frames() callbacks looks different in platform specific implementation.

Thus, I see no dead code here (`GNTTABOP_setup_table` and status check, which are the points of this discussion are performed for both platforms).